sharedbuffers/mapped_struct.py

```diff
@@ -464,29 +505,42 @@ def get_wrapped_key(obj):
     # the key for the unwrapped value
     if isinstance(obj, tuple):
         return obj[1]
-    elif isinstance(obj, int):
-        return obj
-    elif isinstance(obj, long):
-        return obj & ~WRAP_MASK
+    elif python2:
+        if isinstance(obj, int):
+            return obj
+        elif isinstance(obj, long):
+            return obj & ~WRAP_MASK
+    else:
+        if isinstance(obj, int):
+            return obj & ~WRAP_MASK
 
 
 class mapped_frozenset(frozenset):
 
     @classmethod
-    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong)
+    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, cdtype=cython.uchar, udtype="unicode")
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
         if isinstance(obj, npndarray):
             all_int = all_intlong = all_float = 0
             obj_dtype = obj.dtype
             if obj_dtype.isbuiltin:
-                dtype = obj_dtype.char
-                cdtype = cython.cast('const char*', dtype)[0]
-                if cdtype in ('l', 'I', 'i', 'H', 'h', 'B', 'b'):
-                    all_int = all_intlong = 1
-                elif cdtype == 'L':
-                    all_intlong = 1
-                elif cdtype in ('d', 'f'):
-                    all_float = 1
+                if python3:
+                    udtype = obj_dtype.char
+                    if udtype[0] in (u'l', u'I', u'i', u'H', u'h', u'B', u'b'):
+                        all_int = all_intlong = 1
+                    elif udtype[0] == u'L':
+                        all_intlong = 1
+                    elif udtype[0] in (u'd', u'f'):
+                        all_float = 1
```

Comment:
```suggestion
                    udtype = obj_dtype.char
                    udtype0 = udtype[0]
                    if udtype0 in (u'l', u'I', u'i', u'H', u'h', u'B', u'b'):
                        all_int = all_intlong = 1
                    elif udtype0 == u'L':
                        all_intlong = 1
                    elif udtype0 in (u'd', u'f'):
                        all_float = 1
```

Seems to generate substantially better code